petaio.c

```diff
@@ -124,6 +125,106 @@ petaio_build_selection(int * selection,
     }
 }
 
+/*These two functions are only ued for the semi-linear neutrino implementation,
+ * and store data specific to that in the snapshots*/
+/*Defined in gravpm.c*/
+extern _delta_tot_table delta_tot_table;
+
+void petaio_save_neutrinos(BigFile * bf)
+{
+#pragma omp master
+    {
+    double * scalefact = delta_tot_table.scalefact;
+    size_t nk = delta_tot_table.nk, ia = delta_tot_table.ia;
+    size_t ik, i;
+    double * delta_tot = mymalloc("tmp_delta",nk * ia * sizeof(double));
+    /*Save a flat memory block*/
+    for(ik=0;ik< nk;ik++)
+        for(i=0;i< ia;i++)
+            delta_tot[ik*ia+i] = delta_tot_table.delta_tot[ik][i];
+
+    BigBlock bn = {0};
+    if(0 != big_file_mpi_create_block(bf, &bn, "Neutrino", NULL, 0, 0, 0, MPI_COMM_WORLD)) {
+        endrun(0, "Failed to create block at %s:%s\n", "Neutrino",
+                big_file_get_error_message());
+    }
+    if ( (0 != big_block_set_attr(&bn, "Nscale", &ia, "u8", 1)) ||
+       (0 != big_block_set_attr(&bn, "scalefact", scalefact, "f8", ia)) ||
+        (0 != big_block_set_attr(&bn, "Nkval", &nk, "u8", 1)) ) {
+        endrun(0, "Failed to write neutrino attributes %s\n",
+                    big_file_get_error_message());
+    }
+    if(0 != big_block_mpi_close(&bn, MPI_COMM_WORLD)) {
+        endrun(0, "Failed to close block %s\n",
+                    big_file_get_error_message());
+    }
+    BigArray deltas = {0};
+    size_t dims[2] = {0 , 1};
+    /*The neutrino state is shared between all processors,
+     *so only write on master task*/
+    if(ThisTask == 0)
+        dims[0] = nk*ia;
+    ptrdiff_t strides[2] = {8, 8};
+    big_array_init(&deltas, delta_tot, "=f8", 2, dims, strides);
+    petaio_save_block(bf, "Neutrino/deltas", &deltas);
```

Comment:
Perhaps saving k as well?And using captialized column names, Delta, K?

Reply:
Looks like you can set dims[2] to {nk, ia} -- the result will read into python as a 2-d table.

Reply:
Is it ok to do that and read it back into a flat memory block in C?

Reply:
It is still a flat memory block as long as the strides are correctly set.. I think you want
```
strides[2] = {8 * ia, 8}
dims[2] = {nk or 0, ia}
```
This will give you a tight representation of the data in C after reading.

Reply:
Thanks - does what I have now look right?